library/Fission/User/DID/Types.hs

```diff
@@ -1,27 +1,117 @@
+>   public key fingerprint using SHA2-256/256 would look like this:
+> 0x7a 0x5d 0x12 0x20 HASH_BYTES
+https://github.com/w3c-ccg/lds-ed25519-2018/issues/3
+
+> So the procedure to express other types of keys is:
+>
+>    1. Find the multicodec representation for your key type.
+>         For example, the registration for secp256k1 keys is pending
+>         over at multiformats/multicodec#160.
+>    2. Encode the raw public key bytes using that multicodec value.
+>    3. Pick a character encoding (I suggest base58-btc),
+>         and convert it to a character string using Multibase.
+>    4. Prepend did:key: to that string, and there you have it!
+https://github.com/w3c-ccg/did-method-key/issues/3#issuecomment-594190524
+
+NOTE: The multihash is in a middle-endian format -- SB base-7 with continuation flags
+More here: https://github.com/multiformats/unsigned-varint
+
+==== __Examples__
+
+>>> decode' "\"did:key:zBR4m3DNZHT1G8Nb2RHzgKK7TrWxEmJjZskgvFeJwYJ6kpzy1PVDvn3jR2vaAWExNdtKT7KzBoAdy8GHeGd8jpiAUDgbRRnMy\"" :: Maybe DID
+Just (DID {publicKey = Public {publicKey = "AAAAC3NzaC1lZDI1NTE5AAAAIIPnL+R9+OrIm26I1MSOnu4ofAtJ5PjmfiO9ukShjoST"}, algorithm = Ed25519, method = Key})
+
+>>> decode' "\"did:key:z1ArJG77479Lcdo6AYX1efdvVo2dhSeAwPpazVKhFb8Sg4w35AKjx4sdfLy36LnuBq8b3JGt4SJCsnT3mmnujxxehgTiNcXw495k5dafYE9Zt77T8xCWdDkrLLvbGBzV17vNunkgkZKaA5WqwhV4QmFdFCCvEHszNe62SxFCR6g2YqAimX1DSWwVQFc12iuRYrXj8nWeMv2iQG3ZMCvKo2XsW7SXR1Kai29HZdwFQQneEdoEbTWHtFRUh5MpZuX95WNkaZy4JWHjQUNGgNRM6gGrgnmF5ibbXXoTNGhwCH2Qt9DR9khd6t38ufNUdcffapWk5cuGo5pfM1Zye9vapGMpLxQgRUnSyXiz41tSbMHzveTYggiCShVZJ27VLigREDMBqbU8bUNXZBL5L2yTaQieUhPbHsuM1FUD4AGfGQwAmdUstsMC4tjLoYjRBe6vJnwMMxB2ks89M2Yu2aYPf5WgkRGAX9EeXmmiq72A2dPThP4XGaP8pmWw7658BUTuA\"" :: Maybe DID
+Just (DID {publicKey = Public {publicKey = "AAAAB3NzaC1yc2EAAAADAQABAAABAQDkrRwcO9XZOWdwcK9CUQbzD3NMGlmkoRWu/BS5b/C9lm7PIyjBIhshnd6Y29upBKra7dJ7b1qOJDRQS5uvu93OZi/6pGXcqlYHS9WWJtpEQM+VXeJ2PcnKl5ok2mWgeOEqjHRorT+2dVlISjvOk4dRTJR2sB3el8ynQ1W7LuiEio22352O0DYV89DMhMPVVoSvXVBbsvuJv4VJ4e2XYlilsYyF/6zba4rvEP37MJBExNUqlWUbmIAzFbSoJSdickzHJtLCaBu8Eapu/bu90ecNiFIEaXDSvjD+wVqNwqaarWDor248BULN0u3mVTxHh185k8kBAK6ITBnDMJzjsk11"}, algorithm = RSA2048, method = Key})
+
+-}
+data DID = DID
+  { publicKey :: !Key.Public
+  , algorithm :: !Key.Algorithm
+  , method    :: !Method
+  } deriving (Show, Eq)
+
+instance Arbitrary DID where
+  arbitrary = do
+    publicKey <- arbitrary
+    algorithm <- arbitrary
+    method    <- arbitrary
+
+    return DID {..}
+
+instance ToJSON DID where
+  toJSON (DID (Key.Public pk) algo method) =
+    String (header <> UTF8.toBase58Text multicodec64)
+    where
+      header :: Text
+      header = "did:" <> textDisplay method <> ":" <> "z"
+
+      multicodec64 :: ByteString
+      multicodec64 = BS.pack magicBytes <> encodeUtf8 pk
+
+      magicBytes :: [Word8]
+      magicBytes =
+        case algo of
+          Ed25519 -> [0xed, 0x01]
+          RSA2048 -> [0x00, 0x75, 0x01]
```

Comment:
got a different magic number for RSA (373): [0xf5, 0x03] instead of [0x75, 0x01]
0xf5 is the same as 0x75 except for the leading bit, which needs to be a 1 since there is another byte that follows
the second number needs to equal 256. from their examples 128 == 00000001 == 0x01 so 256 == 00000010 == 0x03
lmk if I'm thinking about it wrong

Reply:
Working out loud:

```
From example

11111111 00000001 = 255 
c1111111 c0000001
 |--+--|        ^
    |           |
   127   +     128 = 255 ✅
```

```
  0x75     0x02
01110101 00000010
c1110101 c0000010
 |--+--|       ^
    |          |
   117   +    256 = 373 ✅
```

> 0xf5 is the same as 0x75 except for the leading bit, which needs to be a 1 since there is another byte that follows

Oh god, I did this _again_ in the above 😆

```
  0xF5     0x02
11110101 00000010
```

So we have a discrepancy

`[0xF5, 0x02]` vs `[0xF5, 0x02]`

Can you double check our math again please? 😅

Reply:
oh yeah it's totally 0x02 idk why my brain interpreted that as 0x03.

Yeah so just so we're in agreement:
:guardsman::guardsman:  `[0xF5, 0x02]` :guardsman::guardsman: